File: platform/polyglot/langgraph-flow-serializer/py/langgraph_flow_serializer/deserializer.py
"""
Deserializer: JSON string -> AIWorkflow.

Mirrors mjs/src/deserializer.ts:
  - Parses JSON
  - Normalizes optional fields to defaults
  - Validates via Pydantic schema
  - Raises ValidationError on failure
"""

from __future__ import annotations

import json

from pydantic import ValidationError as PydanticValidationError

from langgraph_flow_serializer.schema import AIWorkflowModel
from langgraph_flow_serializer.types import (
    AIEdge,
    AINode,
    AINodeData,
    AIViewport,
    AIWorkflow,
    ValidationError,
)


def _normalize(raw: dict) -> dict:
    """Apply defaults for missing optional fields before Pydantic validation."""
    normalized = dict(raw)

    normalized.setdefault("nodes", [])
    normalized.setdefault("edges", [])

    # Normalize each node
    nodes = []
    for n in normalized["nodes"]:
        node = dict(n)
        data = dict(node.get("data", {}))
        data.setdefault("inputs", {})
        # Map camelCase JSON keys to snake_case for Pydantic model
        if "nodeType" in data and "node_type" not in data:
            data["node_type"] = data.pop("nodeType")
        node["data"] = data
        nodes.append(node)
    normalized["nodes"] = nodes

    # Normalize each edge
    edges = []
    for e in normalized["edges"]:
        edge = dict(e)
        if "sourceHandle" in edge and "source_handle" not in edge:
            edge["source_handle"] = edge.pop("sourceHandle")
        if "targetHandle" in edge and "target_handle" not in edge:
            edge["target_handle"] = edge.pop("targetHandle")
        edges.append(edge)
    normalized["edges"] = edges

    return normalized


def _model_to_workflow(model: AIWorkflowModel) -> AIWorkflow:
    """Convert a validated Pydantic model to dataclass instances."""

    def to_node(nm) -> AINode:
        data = AINodeData(
            node_type=nm.data.node_type,
            category=nm.data.category,
            name=nm.data.name,
            inputs=nm.data.inputs,
            credentials=nm.data.credentials,
        )
        return AINode(
            id=nm.id,
            type=nm.type,
            position={"x": nm.position.get("x", 0), "y": nm.position.get("y", 0)},
            data=data,
        )

    def to_edge(em) -> AIEdge:
        return AIEdge(
            id=em.id,
            source=em.source,
            source_handle=em.source_handle,
            target=em.target,
            target_handle=em.target_handle,
            type=em.type,
        )

    viewport = None
    if model.viewport is not None:
        viewport = AIViewport(
            x=model.viewport.x,
            y=model.viewport.y,
            zoom=model.viewport.zoom,
        )

    return AIWorkflow(
        id=model.id,
        name=model.name,
        description=model.description,
        viewport=viewport,
        nodes=[to_node(n) for n in model.nodes],
        edges=[to_edge(e) for e in model.edges],
    )


def deserialize(json_str: str) -> AIWorkflow:
    """
    Deserialize a JSON string to an AIWorkflow.

    Args:
        json_str: The JSON string to parse and validate.

    Returns:
        A validated AIWorkflow with defaults applied.

    Raises:
        ValidationError: If the JSON is malformed or fails schema validation.
    """
    try:
        raw = json.loads(json_str)
    except json.JSONDecodeError as exc:
        raise ValidationError(f"Failed to parse JSON: {exc}") from exc

    if not isinstance(raw, dict):
        raise ValidationError("Workflow JSON must be an object")

    normalized = _normalize(raw)

    try:
        model = AIWorkflowModel.model_validate(normalized)
    except PydanticValidationError as exc:
        issues = [f"{'.'.join(str(loc) for loc in e['loc'])}: {e['msg']}" for e in exc.errors()]
        raise ValidationError(
            f"Workflow validation failed:\n  " + "\n  ".join(issues),
            issues=issues,
        ) from exc

    return _model_to_workflow(model)

File: platform/polyglot/langgraph-flow-serializer/mjs/src/deserializer.ts
/**
 * Deserializer: JSON string → AIWorkflow.
 *
 * Parses the JSON, validates it against the Zod schema, normalizes optional
 * fields to sensible defaults, and throws a ValidationError on any failure.
 */

import type { AIWorkflow } from "./types.js";
import { ValidationError } from "./types.js";
import { AIWorkflowSchema } from "./schema.js";
import { ZodError } from "zod";

function normalizeDefaults(raw: unknown): unknown {
  if (typeof raw !== "object" || raw === null) return raw;

  const obj = raw as Record<string, unknown>;

  // Normalize top-level optional fields
  const normalized: Record<string, unknown> = {
    ...obj,
    nodes: Array.isArray(obj.nodes) ? obj.nodes : [],
    edges: Array.isArray(obj.edges) ? obj.edges : [],
  };

  // Normalize each node's data.inputs default
  normalized.nodes = (normalized.nodes as unknown[]).map((n) => {
    if (typeof n !== "object" || n === null) return n;
    const node = n as Record<string, unknown>;
    const data = (typeof node.data === "object" && node.data !== null
      ? node.data
      : {}) as Record<string, unknown>;
    return {
      ...node,
      data: {
        inputs: {},
        ...data,
      },
    };
  });

  return normalized;
}

/**
 * Deserialize a JSON string to an AIWorkflow.
 *
 * @param json - The JSON string to deserialize.
 * @returns A validated AIWorkflow object with defaults applied.
 * @throws ValidationError if parsing or validation fails.
 */
export function deserialize(json: string): AIWorkflow {
  let raw: unknown;

  try {
    raw = JSON.parse(json);
  } catch (err) {
    throw new ValidationError(
      `Failed to parse JSON: ${err instanceof Error ? err.message : String(err)}`
    );
  }

  const withDefaults = normalizeDefaults(raw);

  try {
    return AIWorkflowSchema.parse(withDefaults) as AIWorkflow;
  } catch (err) {
    if (err instanceof ZodError) {
      const issues = err.issues.map(
        (issue) => `${issue.path.join(".") || "root"}: ${issue.message}`
      );
      throw new ValidationError(
        `Workflow validation failed:\n  ${issues.join("\n  ")}`,
        issues
      );
    }
    throw new ValidationError(
      `Workflow validation failed: ${err instanceof Error ? err.message : String(err)}`
    );
  }
}
